hyperjumptech/monika
# Add readfile retries when re-reading configuration file

# Monika Pull Request (PR)  

## What feature/issue does this PR add  
1. Improves on issue #528 
2. When the config file is updated, monika reloads the config
3. But if the file is still being updated, monika may fail

## How did you implement / how did you fix it  
1.  Add a retry loop when reading configfile
2. Add a delay between reads (about 700ms)
3. 700ms delay was chosen after several tweaks and tests

## How to test  

Method 1 : automated
1. Run monika with the config file "testchange.yml" ex: `monika -c testchange.yml -s 0`
4. Then (in another shell) run the config changer loop script : `npx ts-node ./test/others/change-config-loop.ts`
5. You should get an exception "Failed to read testchange.yml. Got empty"

Before the fix: Using change-config-loop script to modify config file:
![image](https://user-images.githubusercontent.com/964106/163938884-0436d192-0185-4cbd-b265-b9d4bffd6d57.png)

Method 2: Manual
1. Run monika with any config file you want
2. Using an editor, while monika is running, change the config file. And Save.

After the fix, after a few thousand iterations, still not glitching:
![image](https://user-images.githubusercontent.com/964106/163939157-e691bfd0-77d9-4a26-9afc-d6887bf3c7f0.png)
 
## Some Testing Notes
1. This seems to be very environment dependent. On my daily machine (Pop_OS 21.10) it rarely happened, even with the loop changer script.
2. On a Windows machine it happened quite easily. I had the sources on WSL, but editing from vscode on the main Windows environment.
3. Updating the config file on vscode seems to take forever on Windows, which results int he 10x loop retry implementation.

Debugging on reload retries. Alternative len a debug output, shows file length and reread attempts.
![image](https://user-images.githubusercontent.com/964106/163940840-e30c0ef9-855a-4667-b18a-49638bc26ad2.png)
